src/Lean/Data/FuzzyMatching.lean

```diff
@@ -0,0 +1,189 @@
+/-
+Copyright (c) 2018 Microsoft Corporation. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Author: Lars König
+
+The string fuzzy matching algorithm in this file is based on the algorithm
+used in LLVM with some modifications. The LLVM algorithm itself is based on VS
+code's client side filtering algorithm.
+-/
+
+namespace Lean
+namespace FuzzyMatching
+
+/- Represents the type of a single character. -/
+inductive CharType where
+  | lower | upper | separator
+
+def charType (c : Char) : CharType :=
+  if c.isAlphanum then
+    if c.isUpper
+      then CharType.upper
+      else CharType.lower
+  else
+    CharType.separator
+
+
+/- Represents the role of a character inside a word. -/
+inductive CharRole where
+  | head | tail | separator
+
+def charRole (prev? : Option CharType) (curr : CharType) : CharRole :=
+  if curr matches CharType.separator then
+    CharRole.separator
+  else if prev? matches none || prev? matches some CharType.separator then
+    CharRole.head
+  else if curr matches CharType.lower then
+    CharRole.tail
+  else if prev? matches some CharType.upper then
+    CharRole.tail
+  else
+    CharRole.head
+
+
+/- Combines different information about a character. -/
+private structure CharInfo where
+  char : Char
+  charLower : Char
+  type : CharType
+  role : CharRole
+
+/- Add additional information to each character in a string and return the
+resulting list in reverse order. -/
+private def reverseStringInfo (s : String) : List CharInfo :=
+  s.toList.foldl (init := (none, [])) (fun (prev, res) curr => (
```

Comment:
From a quick look, this first iterates over the UTF-8 string data to create a list of Unicode codepoints and then folds that. How does the perf compare with a direct iteration like `containsInOrder`? More generally, have you measured where the bottleneck is? The `fuzzyMatchRec` logic or string operations?

Reply:
I'm still fairly new to programming in Lean. Is there any documentation on how we do benchmarking?

Reply:
Not that I am aware of, it is kind of arcane knowledge. But if you know how to build a binary from a Lean file containing `main : IO Unit`, you can benchmark that with `perf` on Linux and view the results in Firefox Profiler ─ see [here](https://github.com/firefox-devtools/profiler/blob/main/docs-user/guide-perf-profiling.md). Lean's mangling scheme appends various internal postfixes to function names in generated code but the results can reasonably be correlated with parts of the original source code.

Reply:
Thank you, I will try this :)

Reply:
I agree with Wojciech that this `List CharInfo` looks like a big potential performance issue.  The other obvious performance hazard is the repeated computation of `toLower` in the main function.

Regarding benchmarking:
1. Please test this using a large data set.  Feel free to grab the list of mathlib declarations: https://leanprover-community.github.io/mathlib_docs/decl.bmp
2. Benchmark queries of different sizes, from 1-2 characters to full names.
3. Make sure that you benchmark the compiled code (function names that show up in the `perf` output are compiled).
4. Just measuring the runtime for the whole search is already an interesting data point.
5. For detailed profiling results, look at `perf` as Wojciech has said.